### PR TITLE
Misc pointer events fixes

### PR DESCRIPTION
**Pull request by dr1rrb**
GitHub Issues:
fixes https://github.com/unoplatform/uno/issues/2193
fixes https://github.com/unoplatform/uno/issues/2180

## Bug fixes
Misc bugs fixes for pointer events

## What is the current behavior?
* On Android, the location of the Tapped event args is always `0,0`
* When in a `ScrollViewer`, the `ToggleSwitch` might stay in _pressed_ visual state
* When 2 nested controls subscribes to the `Tapped` event, the parent one will receive the `Tapped` twice when tapping on the child.
* On WASM the pressed buttons are reported only in `PointerPressed` and `PointerReleased` event. As we validate the pressed state in the `GestureRecognizer`, this means that no gesture can be detected (i.e. `Tapped` and `DoubleTapped` events are never fired)
* The `Thumb` does fire the `DragCompleted` if the event `PointerCanceled` is raised (eg. when scrolling)
* If a child control handles the pointer events, a parent control won't get them so it won't be able to use them with its own `GestureRecognizer` to detect **Manipulation** and **Gesture** (i.e. it won't be able to fire the `Tapped` event for instance).

## What is the new behavior?
_the opposite_

## PR Checklist
- [x] Tested code with current [supported SDKs](https://github.com/unoplatform/uno/blob/master/README.md#uno-features)
- [ ] ~~Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)~~
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [x] [Wasm UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences. Validate PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [ ] ~~Updated the [Release Notes](https://github.com/unoplatform/uno/tree/master/doc/ReleaseNotes)~~
- [x] Associated with an issue (GitHub or internal)

## Internal
Might fix https://nventive.visualstudio.com/_workitems/edit/168971

### CODE REVIEW

File: doc/articles/features/routed-events.md [hunk 1]
```diff
@@ -236,6 +236,9 @@ As those events are tightly coupled to the native events, Uno has to make some c
   > a `PointerReleased` when clicking on an `Hyperlink`.
 * Unlike UWP, on the `Hyperlink` the `Click` will be raised before the `PointerReleased`.
 * The property `PointerPointProperties.PointerUpdateKind` is not set on Android 5.x and lower (API level < 23)
+* On Firefox, pressed pointers are reported as finger. This means you will receive events with `PointerDeviceType == Pen` only for hoverring
```
Comment (jeromelaban) [inline doc/articles/features/routed-events.md]:
```suggestion
* On Firefox, pressed pointers are reported as finger. This means you will receive events with `PointerDeviceType == Pen` only for hovering 
```

File: src/Uno.UI/UI/Xaml/Controls/ToggleSwitch.cs [hunk 2]
```diff
@@ -60,11 +56,9 @@ protected override void OnUnloaded()
...
 			if (!IsNativeTemplate)
 			{
-				RemoveHandler(PointerPressedEvent, (PointerEventHandler) OnPointerPressed);
-				RemoveHandler(PointerExitedEvent, (PointerEventHandler) OnPointerExited);
-				RemoveHandler(PointerReleasedEvent, (PointerEventHandler) OnPointerReleased);
-				RemoveHandler(PointerCanceledEvent, (PointerEventHandler) OnPointerCanceled);
-				RemoveHandler(PointerEnteredEvent, (PointerEventHandler) OnPointerEntered);
+				RemoveHandler(PointerPressedEvent, (PointerEventHandler)OnPointerPressed);
+				RemoveHandler(PointerReleasedEvent, (PointerEventHandler)OnPointerReleased);
+				RemoveHandler(PointerCanceledEvent, (PointerEventHandler)OnPointerCanceled);
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/ToggleSwitch.cs]: No more Entered/Exited ?

File: src/Uno.UI/UI/Xaml/Controls/ToggleSwitch.cs [hunk 3]
```diff
@@ -111,15 +92,10 @@ private void OnPointerReleased(object sender, PointerRoutedEventArgs args)
...
-			IsPointerOver = false;
-			IsPointerPressed = false;
-			UpdateCommonState();
-		}
-
-		private void OnPointerEntered(object sender, PointerRoutedEventArgs args)
-		{
-			IsPointerOver = true;
-			UpdateCommonState();
+			if (_switchThumb == null)
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/ToggleSwitch.cs]: What if there is a thumb?

File: src/Uno.UI/UI/Xaml/RoutedEvent.cs [hunk 4]
```diff
@@ -26,14 +22,34 @@ public RoutedEvent([CallerMemberName] string name = null)
...
+		/// <remarks>
+		/// For some events like PointerEvent, we always needs to get the full events sequence to maintain the
+		/// internal state, so we always needs the handled events too.
+		/// This flag avoids the complex update of the SubscribedToHandledEventsToo property coercing and inheritance
+		/// for those kind of well-known events.
+		/// </remarks>
+		[Pure]
+		internal bool IsAlwaysBubbled => IsPointerEvent;
+
+		[Pure]
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/RoutedEvent.cs]: Why adding this now ? What checks it ?

File: src/Uno.UI/UI/Xaml/UIElement.Pointers.wasm.cs [hunk 5]
```diff
@@ -131,50 +118,28 @@ private static PointerRoutedEventArgs PayloadToPointerArgs(object snd, string pa
...
 			var ctrl = parts[3] == "1";
 			var shift = parts[4] == "1";
-			var button = int.Parse(parts[5], CultureInfo.InvariantCulture); // -1: none, 0:main, 1:middle, 2:other (commonly main=left, other=right)
-			var typeStr = parts[6];
-			var srcHandle = int.Parse(parts[7]);
-			var timestamp = double.Parse(parts[8], CultureInfo.InvariantCulture);
+			var buttons = int.Parse(parts[5], CultureInfo.InvariantCulture);
+			var buttonUpdate = int.Parse(parts[6], CultureInfo.InvariantCulture);
+			var typeStr = parts[7];
+			var srcHandle = int.Parse(parts[8]);
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/UIElement.Pointers.wasm.cs]:
```suggestion
			var srcHandle = int.Parse(parts[8], CultureInfo.InvariantCulture);
```

File: src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs [hunk 6]
```diff
@@ -395,7 +397,8 @@ public void RemoveHandler(RoutedEvent routedEvent, object handler)
 				{
 					handlers.Remove(matchingHandler);
 
-					if (matchingHandler.HandledEventsToo)
+					if (matchingHandler.HandledEventsToo
+						&& !routedEvent.IsAlwaysBubbled) // This event is always bubbled, no needs to update the flag
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs]:
```suggestion
						&& !routedEvent.IsAlwaysBubbled) // This event is always bubbled, no need to update the flag
```

File: src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs [hunk 7]
```diff
@@ -450,6 +453,12 @@ private void UpdateSubscribedToHandledEventsToo()
 
 			foreach (var eventHandlers in _eventHandlerStore)
 			{
+				if (eventHandlers.Key.IsAlwaysBubbled)
+				{
+					// This event is always bubbled, no needs to include it in the SubscribedToHandledEventsToo
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs]:
```suggestion
					// This event is always bubbled, no need to include it in the SubscribedToHandledEventsToo
```

File: src/Uno.UWP/UI/Input/GestureRecognizer.cs [hunk 8]
```diff
@@ -179,8 +179,8 @@ private void Recognize(List<PointerPoint> points, PointerPoint pointerUp)
 
 		#region Tap (includes DoubleTap)
 		internal const ulong MultiTapMaxDelayTicks = TimeSpan.TicksPerMillisecond * 1000;
-		internal const int TapMaxXDelta = 15;
-		internal const int TapMaxYDelta = 15;
+		internal const int TapMaxXDelta = 10;
```
Comment (jeromelaban) [inline src/Uno.UWP/UI/Input/GestureRecognizer.cs]: Why ? (why previous, why new)

File: src/Uno.UI/UI/Xaml/Controls/ToggleSwitch.cs (hunk 2 above)
Comment (dr1rrb) [reply src/Uno.UI/UI/Xaml/Controls/ToggleSwitch.cs]: No, there were used only for visual states, which are now handled by the mixins

File: src/Uno.UI/UI/Xaml/Controls/ToggleSwitch.cs (hunk 3 above)
Comment (dr1rrb) [reply src/Uno.UI/UI/Xaml/Controls/ToggleSwitch.cs]: We will get a `DragCompleted` instead.

File: src/Uno.UWP/UI/Input/GestureRecognizer.cs (hunk 8 above)
Comment (dr1rrb) [reply src/Uno.UWP/UI/Input/GestureRecognizer.cs]: Previous : Random
New: Still random ... but found that the delta was a bit too high to use it properly in the `ToggleSwitch`. I've did a commit just for this change in order to be able to track it a bit more easily : https://github.com/unoplatform/uno/pull/2249/commits/93c6b674afa2b1ed69212ec804711558774c26f3

File: src/Uno.UI/UI/Xaml/RoutedEvent.cs (hunk 4 above)
Comment (dr1rrb) [reply src/Uno.UI/UI/Xaml/RoutedEvent.cs]: At the beginning I add it only on the `IsAlwaysBubbled` as this flag is used in code that might run quite frequently, I wanted to hint the compilator. But I agree that I'm not sure if the compilator use it, nor if it's needed on auto-properties.

BTW: ReSharper checks it.

File: src/Uno.UI/UI/Xaml/Input/DoubleTappedRoutedEventArgs.cs [hunk 9]
```diff
@@ -10,33 +10,36 @@ namespace Windows.UI.Xaml.Input
...
+				return default; // Required for the default public ctor ...
+			}
+			else if(relativeTo == _originalSource)
 			{
 				return _position;
 			}
 			else
 			{
-				return relativeTo.GetPosition(_position, relativeTo);
+				return _originalSource.GetPosition(_position, relativeTo);
```
Comment (carldebilly) [inline src/Uno.UI/UI/Xaml/Input/DoubleTappedRoutedEventArgs.cs]: So `relativeTo` is not used anymore?

Comment (dr1rrb) [reply src/Uno.UI/UI/Xaml/Input/DoubleTappedRoutedEventArgs.cs]: It is ... at the end of the line ... that was the issue!

File: doc/articles/features/routed-events.md [hunk 10]
```diff
@@ -236,6 +236,9 @@ As those events are tightly coupled to the native events, Uno has to make some c
   > a `PointerReleased` when clicking on an `Hyperlink`.
 * Unlike UWP, on the `Hyperlink` the `Click` will be raised before the `PointerReleased`.
 * The property `PointerPointProperties.PointerUpdateKind` is not set on Android 5.x and lower (API level < 23)
+* On Firefox, pressed pointers are reported as finger. This means you will receive events with `PointerDeviceType == Pen` only for hovering
```
Comment (jeromelaban) [inline doc/articles/features/routed-events.md]:
```suggestion
* On Firefox, pressed pointers are reported as fingers. This means you will receive events with `PointerDeviceType == Pen` only for hovering 
```

File: src/Uno.UI/UI/Xaml/Controls/ToggleSwitch.cs [hunk 11]
```diff
@@ -259,21 +245,26 @@ private void OnDragDelta(object sender, DragDeltaEventArgs e)
 
 		private void OnDragCompleted(object sender, DragCompletedEventArgs e)
 		{
-			if (InterpretAsTap)
+			// If the user only drags the thumb by a few pixels before releasing it, 
+			// we interpret it as a Tap rather than a drag gesture.
+			// Note: We do not use the Tapped event as this offers a better sync between
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/Controls/ToggleSwitch.cs]: between what? Tell us!

Comment (dr1rrb) [reply src/Uno.UI/UI/Xaml/Controls/ToggleSwitch.cs]: shhhhh it's a secret 🤫

**Pull request merged**